cmd/root.go

```diff
@@ -529,30 +511,169 @@ status code.`)
+	c.SilenceUsage = true
+
+	// Errors will be handled by logging from here on.
+	c.SilenceErrors = true
+
+	return nil
+}
+
+func getViper(c *Command) (*viper.Viper, error) {
+	v := viper.New()
+
+	if c.conf.ConfigFilepath != "" {
+		// Setup Viper configuration file. Viper will attempt to load configuration
+		// from the specified file if it exists. Otherwise, Viper will source all
+		// configuration from flags and then environment variables.
+		ext := filepath.Ext(c.conf.ConfigFilepath)
+
+		if ext == "" {
+			return nil, newBadCommandError(fmt.Sprintf("config file %v has no extension, "+
+				"requires TOML, YAML or JSON format", c.conf.ConfigFilepath))
 		}
-	})
 
-	return c
+		if ext != ".toml" && ext != ".yaml" && ext != ".yml" && ext != ".json" {
+			return nil, newBadCommandError(fmt.Sprintf("config file %v has an invalid extension, "+
+				"requires TOML, YAML or JSON format", c.conf.ConfigFilepath))
+		}
+
+		conf := filepath.Base(c.conf.ConfigFilepath)
+		noExt := strings.ReplaceAll(conf, ext, "")
+		v.SetConfigName(noExt)   // argument must be the name of config file without extension
+		v.SetConfigType(ext[1:]) // setting the config type makes the extension optional
+		v.AddConfigPath(filepath.Dir(c.conf.ConfigFilepath))
+
+		// Attempt to load configuration from a file. If no file is found, assume
+		// configuration is provided by flags or environment variables.
+		if err := v.ReadInConfig(); err != nil {
+			// If the error is a ConfigFileNotFoundError, then ignore it.
+			// Otherwise, report the error to the user.
+			var configFileNotFoundError viper.ConfigFileNotFoundError
+			if !errors.As(err, &configFileNotFoundError) {
+				// TODO clean up this error message
+				return nil, newBadCommandError(fmt.Sprintf(
+					"failed to load configuration from %v: %v",
+					c.conf.ConfigFilepath, err,
+				))
+			}
+		}
+	}
+
+	v.SetEnvPrefix(envPrefix)
+	v.SetEnvKeyReplacer(strings.NewReplacer("-", "_"))
+	v.AutomaticEnv()
+
+	return v, nil
+}
+
+func instanceFromEnv(args []string) []string {
+	// This supports naming the first instance first with:
+	//     INSTANCE_CONNECTION_NAME
+	// or if that's not defined, with:
+	//     INSTANCE_CONNECTION_NAME_0
+	inst := os.Getenv(fmt.Sprintf("%s_INSTANCE_CONNECTION_NAME", envPrefix))
+	if inst == "" {
+		inst = os.Getenv(fmt.Sprintf("%s_INSTANCE_CONNECTION_NAME_0", envPrefix))
+		if inst == "" {
+			return nil
+		}
+	}
+	args = append(args, inst)
+
+	i := 1
+	for {
+		instN := os.Getenv(fmt.Sprintf("%s_INSTANCE_CONNECTION_NAME_%d", envPrefix, i))
+		// if the next instance connection name is not defined, stop checking
+		// environment variables.
+		if instN == "" {
+			break
+		}
+		args = append(args, instN)
+		i++
+	}
+	return args
+}
+
+func instanceFromConfigFile(v *viper.Viper) []string {
+	var args []string
+	inst := v.GetString("instance_connection_name")
+
+	if inst == "" {
+		inst = v.GetString("instance_connection_name_0")
```

Comment:
Let's add a test for this behavior, i.e., config file with:

``` toml
instance_connection_name_0 = x:y:z
instance_connection_name_1 = a:b:c
```

Reply:
added a test for this 🙂